File: adf_core_python/core/component/gateway/message/am_update.py
from abc import ABC
from typing import Any

from rcrs_core.connection import RCRSProto_pb2
from rcrs_core.messages.message import Message
from rcrs_core.worldmodel.changeSet import ChangeSet
from rcrs_core.worldmodel.entityID import EntityID

from adf_core_python.core.component.gateway.message.urn.urn import (
    ModuleMSG,
    ComponentModuleMSG,
)


class AMUpdate(Message, ABC):
    def __init__(self) -> None:
        super().__init__(ModuleMSG.AM_UPDATE)

    def read(self) -> None:
        pass

    def write(self, agent_id: EntityID, time: int, changed: ChangeSet, heard) -> Any:
        msg = RCRSProto_pb2.MessageProto()
        msg.urn = self.get_urn()
        msg.components[ComponentModuleMSG.AgentID].entityID = agent_id.get_value()
        msg.components[ComponentModuleMSG.Time].intValue = time
        msg.components[ComponentModuleMSG.Changed].changeSet = changed
        msg.components[ComponentModuleMSG.Heard].commandList = heard

        return msg

File: java/lib/src/main/java/adf_core_python/gateway/message/AMUpdate.java
package adf_core_python.gateway.message;

import adf_core_python.gateway.message.urn.ModuleMessageComponentURN;
import adf_core_python.gateway.message.urn.ModuleMessageURN;
import rescuecore2.messages.AbstractMessage;
import rescuecore2.messages.Command;
import rescuecore2.messages.components.ChangeSetComponent;
import rescuecore2.messages.components.CommandListComponent;
import rescuecore2.messages.components.IntComponent;
import rescuecore2.messages.protobuf.RCRSProto;
import rescuecore2.worldmodel.ChangeSet;

import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;
import java.util.List;

public class AMUpdate extends AbstractMessage {
    private final IntComponent time;
    private final ChangeSetComponent changed;
    private final CommandListComponent heard;

    public AMUpdate(int time, ChangeSet changed, Collection<Command> heard) {
        this();
        this.time.setValue(time);
        this.changed.setChangeSet(changed);
        this.heard.setCommands(heard);
    }

    public AMUpdate(InputStream inputStream) throws IOException {
        this();
        read(inputStream);
    }

    public AMUpdate(RCRSProto.MessageProto messageProto) {
        this();
        fromMessageProto(messageProto);
    }

    private AMUpdate() {
        super(ModuleMessageURN.AM_UPDATE);
        this.time = new IntComponent(ModuleMessageComponentURN.Time);
        this.changed = new ChangeSetComponent(ModuleMessageComponentURN.Changed);
        this.heard = new CommandListComponent(ModuleMessageComponentURN.Heard);
        addMessageComponent(this.time);
        addMessageComponent(this.changed);
        addMessageComponent(this.heard);
    }

    public int getTime() {
        return this.time.getValue();
    }

    public ChangeSet getChanged() {
        return this.changed.getChangeSet();
    }

    public List<Command> getHeard() {
        return this.heard.getCommands();
    }
}
